### ISSUE TITLE
Reverse portfwd looks broken

### ISSUE THREAD

**Issue by plowsec** [rapid7/metasploit-framework#15968]
## Steps to reproduce

How'd you do it?

1. Host machine is macOS or Linux with latest metasploit-framework. Victim machine is a Windows 10. Payloads used are windows/x64/meterpreter/reverse_http

1. `use exploit/multi/handler` with config `lhost=192.168.18.1` and `lport=80`
2. spawn first shell with a reverse_http payload -> works
3. `portfwd add -R -r 127.0.0.1 -l 80 -p 1337 -L 192.168.18.1`
4. run `netstat /a | findstr 1337` on victim -> port is open
5. open web browser on victim machine and browse to "http://127.0.0.1:1337" and you get "It works!" -> ok.
6. generate a reverse_http with config lhost=127.0.0.1 and lport=1337
7. run the payload -> KO 

## Expected behavior

A second session should open

## Current behavior

```
2021-12-16 14:16:51 %2fqueries.json\toto@127.0.0.1 (x64) > portfwd add -R -r 127.0.0.1 -l 80 -p 1337 -L 192.168.18.1
[*] Local TCP relay created: 192.168.18.1:80 <-> 127.0.0.1:1337
2021-12-16 14:16:53 %2fqueries.json\toto@127.0.0.1 (x64) >
[!] http://192.168.18.1:80 handling request from 192.168.18.1; (UUID: wwp2ayn0) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.18.1:80 handling request from 192.168.18.1; (UUID: wwp2ayn0) Staging x64 payload (201308 bytes) ...
[!] http://192.168.18.1:80 handling request from 192.168.18.1; (UUID: wwp2ayn0) Without a database connected that payload UUID tracking will not work!
[-] Meterpreter session 2 is not valid and will be closed
[*]  - Meterpreter session 2 closed.
```

### Metasploit version

2021-12-16 14:27:06 S:1 J:1 exploit(multi/handler) > version
Framework: 6.1.20-dev-4791f70ce3
Console  : 6.1.20-dev-4791f70ce3
```

## Additional Information

```

##  Module/Datastore

The following global/module datastore, and database setup was configured before the issue occurred:
<details>
<summary>Collapse</summary>

```
[framework/core]
MeterpreterUserAgent=Mozilla/5.0 (X11; Linux x86_64; rv:63.0) Gecko/20100101 Firefox/63.0
MeterpreterServerName=Microsoft-IIS/8.0
SMBNAME=LOCALHOST
MeterpreterPrompt=%T %U@%I (%A)
PROMPT=%T S:%S J:%J

[framework/database/local_db_service]
url=[Filtered]

[framework/database/m0qm8L7n]
url=[Filtered]

[framework/features]

[framework/ui/console]
ActiveModule=exploit/multi/handler

[multi/handler]
PAYLOAD=windows/x64/meterpreter/reverse_http
WORKSPACE=
VERBOSE=true
WfsDelay=2
EnableContextEncoding=false
ContextInformationFile=
DisablePayloadHandler=false
ExitOnSession=false
ListenerTimeout=0
lhost=192.168.18.1
lport=80
ReverseListenerBindPort=
ReverseAllowProxy=false
ReverseListenerComm=
PingbackRetries=0
PingbackSleep=30
LURI=
ReverseListenerBindAddress=
OverrideRequestHost=false
OverrideLHOST=
OverrideLPORT=
OverrideScheme=
HttpUserAgent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.69 Safari/537.36
HttpServerName=Apache
HttpUnknownRequestResponse=<html><body><h1>It works!</h1></body></html>
IgnoreUnknownPayloads=false
PayloadUUIDSeed=
PayloadUUIDRaw=
PayloadUUIDName=
PayloadUUIDTracking=false
EnableStageEncoding=false
StageEncoder=
StageEncoderSaveRegisters=
StageEncodingFallback=true
PrependMigrate=false
PrependMigrateProc=
EXITFUNC=process
StagerURILength=
StagerRetryCount=10
StagerRetryWait=5
HttpHostHeader=
HttpCookie=
HttpReferer=
HttpProxyHost=
HttpProxyPort=
HttpProxyUser=
HttpProxyPass=
HttpProxyType=HTTP
AutoLoadStdapi=true
AutoVerifySessionTimeout=30
InitialAutoRunScript=
AutoRunScript=
AutoSystemInfo=true
EnableUnicodeEncoding=false
HandlerSSLCert=
SessionRetryTotal=3600
SessionRetryWait=10
SessionExpirationTimeout=604800
SessionCommunicationTimeout=300
PayloadProcessCommandLine=
AutoUnhookProcess=false
loglevel=3
```

</details>

##  Database Configuration

The database contains the following information:
<details>
<summary>Collapse</summary>

```
Session Type: postgresql selected, no connection
```

</details>

##  History

The following commands were ran during the session and before this issue occurred:
<details>
<summary>Collapse</summary>

```
2000   git log -1 --pretty=oneline
2001   version
2002   set loglevel 3
2003   debug
```

</details>

##  Framework Errors

The following framework errors occurred before the issue occurred:
<details>
<summary>Collapse</summary>

```
[12/16/2021 14:03:54] [e(0)] core: Thread Exception: StreamMonitorRemote  critical=false    source:
    /Users/vladimir/Tools/metasploit-framework/lib/metasploit/framework/thread_factory_provider.rb:25:in `spawn'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/thread_factory.rb:22:in `spawn'
    /Library/Ruby/Gems/2.6.0/gems/rex-core-0.1.17/lib/rex/io/socket_abstraction.rb:120:in `monitor_rsock'
    /Library/Ruby/Gems/2.6.0/gems/rex-core-0.1.17/lib/rex/io/stream_abstraction.rb:30:in `initialize_abstraction'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/channels/socket_abstraction.rb:56:in `initialize'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_client_channel.rb:73:in `initialize'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_server_channel.rb:55:in `new'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_server_channel.rb:55:in `request_handler'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:607:in `block in dispatch_inbound_packet'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:599:in `each'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:599:in `dispatch_inbound_packet'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:698:in `on_passive_request'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:653:in `block in initialize_passive_dispatcher'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/proto/http/handler/proc.rb:38:in `on_request'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/proto/http/server.rb:367:in `dispatch_request'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/proto/http/server.rb:301:in `on_client_data'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/proto/http/server.rb:160:in `block in start'
    /Library/Ruby/Gems/2.6.0/gems/rex-core-0.1.17/lib/rex/io/stream_server.rb:48:in `on_client_data'
    /Library/Ruby/Gems/2.6.0/gems/rex-core-0.1.17/lib/rex/io/stream_server.rb:199:in `block in monitor_clients'
    /Library/Ruby/Gems/2.6.0/gems/rex-core-0.1.17/lib/rex/io/stream_server.rb:197:in `each'
    /Library/Ruby/Gems/2.6.0/gems/rex-core-0.1.17/lib/rex/io/stream_server.rb:197:in `monitor_clients'
    /Library/Ruby/Gems/2.6.0/gems/rex-core-0.1.17/lib/rex/io/stream_server.rb:73:in `block in start'
    /Users/vladimir/Tools/metasploit-framework/lib/rex/thread_factory.rb:22:in `block in spawn'
    /Users/vladimir/Tools/metasploit-framework/lib/msf/core/thread_manager.rb:105:in `block in spawn' - Rex::Post::Meterpreter::RequestError stdapi_net_socket_tcp_shutdown: Operation failed: The handle is invalid.
[12/16/2021 14:15:45] [e(0)] core: Failed to connect to the database: No database YAML file
[12/16/2021 14:16:40] [e(0)] core: Session with no session_host/target_host/tunnel_peer. Session Info: #<Session:meterpreter 127.0.0.1 () >
[12/16/2021 14:16:59] [e(0)] rex: Error in #<Rex::Services::LocalRelay:0x00007ff8bb81f068> monitor_relays read - EOFError end of file reached
[12/16/2021 14:17:29] [e(0)] core: Session with no session_host/target_host/tunnel_peer. Session Info: #<Session:meterpreter 127.0.0.1 () >
[12/16/2021 14:24:02] [e(0)] rex: Error in #<Rex::Services::LocalRelay:0x00007ff8bb81f068> monitor_relays read - EOFError end of file reached
[12/16/2021 14:24:05] [e(0)] rex: Error in #<Rex::Services::LocalRelay:0x00007ff8bb81f068> monitor_relays read - EOFError end of file reached
[12/16/2021 14:27:03] [e(0)] meterpreter: Error running command portfwd: Interrupt
[12/16/2021 14:28:24] [e(0)] rex: Error in #<Rex::Services::LocalRelay:0x00007ff8bb81f068> monitor_relays read - EOFError end of file reached
[12/16/2021 14:28:54] [e(0)] core: Session with no session_host/target_host/tunnel_peer. Session Info: #<Session:meterpreter 127.0.0.1 () >
```

</details>

##  Web Service Errors

The following web service errors occurred before the issue occurred:
<details>
<summary>Collapse</summary>

```
msf-ws.log does not exist.
```

</details>

##  Framework Logs

The following framework logs were recorded before the issue occurred:
<details>
<summary>Collapse</summary>

```
[12/16/2021 14:16:40] [e(0)] core: Session with no session_host/target_host/tunnel_peer. Session Info: #<Session:meterpreter 127.0.0.1 () >
[12/16/2021 14:16:51] [d(0)] core: HistoryManager.push_context name: :meterpreter
[12/16/2021 14:16:59] [e(0)] rex: Error in #<Rex::Services::LocalRelay:0x00007ff8bb81f068> monitor_relays read - EOFError end of file reached
[12/16/2021 14:17:00] [w(0)] core: monitor_rsock: the remote socket is nil, exiting loop
[12/16/2021 14:17:13] [d(0)] core: HistoryManager.pop_context name: :meterpreter
[12/16/2021 14:17:29] [d(0)] core: Session 2 failed to negotiate TLV encryption
[12/16/2021 14:17:29] [e(0)] core: Session with no session_host/target_host/tunnel_peer. Session Info: #<Session:meterpreter 127.0.0.1 () >
[12/16/2021 14:23:22] [d(0)] core: HistoryManager.push_context name: :meterpreter
[12/16/2021 14:24:02] [e(0)] rex: Error in #<Rex::Services::LocalRelay:0x00007ff8bb81f068> monitor_relays read - EOFError end of file reached
[12/16/2021 14:24:05] [w(0)] core: monitor_rsock: the remote socket is nil, exiting loop
[12/16/2021 14:24:05] [e(0)] rex: Error in #<Rex::Services::LocalRelay:0x00007ff8bb81f068> monitor_relays read - EOFError end of file reached
[12/16/2021 14:24:05] [w(0)] core: monitor_rsock: the remote socket is nil, exiting loop
[12/16/2021 14:27:03] [e(0)] meterpreter: Error running command portfwd: Interrupt
[12/16/2021 14:27:03] [d(0)] meterpreter: Call stack:
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:95:in `sleep'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:95:in `wait'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:95:in `block in wait'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:93:in `synchronize'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_response_waiter.rb:93:in `wait'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:234:in `send_packet_wait_response'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:176:in `send_request'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/channel.rb:116:in `create'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_server_channel.rb:72:in `open'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket.rb:107:in `create_tcp_server_channel'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket.rb:91:in `create'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/net.rb:454:in `cmd_portfwd'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:563:in `run_command'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/ui/console.rb:102:in `run_command'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:512:in `block in run_single'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:506:in `each'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:506:in `run_single'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/ui/console.rb:64:in `block in interact'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/shell.rb:157:in `run'
/Users/vladimir/Tools/metasploit-framework/lib/rex/post/meterpreter/ui/console.rb:62:in `interact'
/Users/vladimir/Tools/metasploit-framework/lib/msf/base/sessions/meterpreter.rb:555:in `_interact'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/interactive.rb:53:in `interact'
/Users/vladimir/Tools/metasploit-framework/lib/msf/ui/console/command_dispatcher/core.rb:1591:in `cmd_sessions'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:563:in `run_command'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:512:in `block in run_single'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:506:in `each'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:506:in `run_single'
/Users/vladimir/Tools/metasploit-framework/lib/rex/ui/text/shell.rb:162:in `run'
/Users/vladimir/Tools/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/Users/vladimir/Tools/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
./msfconsole:23:in `<main>'
[12/16/2021 14:27:04] [d(0)] core: HistoryManager.pop_context name: :meterpreter
[12/16/2021 14:28:24] [e(0)] rex: Error in #<Rex::Services::LocalRelay:0x00007ff8bb81f068> monitor_relays read - EOFError end of file reached
[12/16/2021 14:28:28] [w(0)] core: monitor_rsock: the remote socket is nil, exiting loop
[12/16/2021 14:28:54] [d(0)] core: Session 3 failed to negotiate TLV encryption
[12/16/2021 14:28:54] [e(0)] core: Session with no session_host/target_host/tunnel_peer. Session Info: #<Session:meterpreter 127.0.0.1 () >
```

**Comment by sempervictus** [comment]
Ping @OJ - is this on the meterp end, in the packet handler, or at the actual socket monitor? :smile: This smells eerily familiar.

**Comment by smcintyre-r7** [comment]
Well I was able to reproduce this. I suspect the issue is related to the reverse_http payload and or listener. I was able to get everything working just fine with reverse_tcp.

FWIW you can skip the explicit reverse portfwd and just use the `ReverseListenerComm` advanced argument. Set it to the session that you want used to create the listener (the session where you would have run the portfwd from). That will also help Metasploit know what session the connection is pivoted through.

**Comment by smcintyre-r7** [comment]
I've made some substantial progress on this. The changes are pretty invasive so I'll want to test it a bit more thoroughly but whereas before it was failing 10/10 times I am now seeing it work 10/10 times. I simplified the steps to pivoting a reverse_http Python Meterpreter through a reverse_tcp Windows meterpreter though because it was easier to debug. I'll retest the original steps as well to ensure the issues I addressed were the ones that were originally identified.
